pkg/controller/atlasstream/atlasstream_connection.go

```diff
@@ -0,0 +1,256 @@
+			ops.Update = append(ops.Update, &akoConnection)
+
+			continue
+		}
+
+		ops.NoOp = append(ops.NoOp, &akoConnection)
+	}
+
+	for i := range atlasStreamConnections {
+		if _, ok := akoConnectionByName[atlasStreamConnections[i].GetName()]; !ok {
+			ops.Delete = append(ops.Delete, &atlasStreamConnections[i])
+		}
+	}
+
+	return ops, nil
+}
+
+func streamConnectionToAtlas(ctx context.Context, k8sClient client.Client) streamConnectionMapper {
+	return func(streamConnection *akov2.AtlasStreamConnection) (*admin.StreamsConnection, error) {
+		connection := admin.StreamsConnection{
+			Name: &streamConnection.Spec.Name,
+			Type: &streamConnection.Spec.ConnectionType,
+		}
+
+		if streamConnection.Spec.ClusterConfig != nil {
+			connection.ClusterName = &streamConnection.Spec.ClusterConfig.Name
+			connection.DbRoleToExecute = &admin.DBRoleToExecute{
+				Role: &streamConnection.Spec.ClusterConfig.Role.Name,
+				Type: &streamConnection.Spec.ClusterConfig.Role.RoleType,
+			}
+		}
+
+		if streamConnection.Spec.KafkaConfig != nil {
+			authData, err := getSecretData(
+				ctx,
+				k8sClient,
+				*streamConnection.Spec.KafkaConfig.Authentication.Credentials.GetObject(streamConnection.Namespace),
+				kafkaConnectionAuthUsername,
+				kafkaConnectionAuthPassword,
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			secData, err := getSecretData(
+				ctx,
+				k8sClient,
+				*streamConnection.Spec.KafkaConfig.Security.Certificate.GetObject(streamConnection.Namespace),
+				kafkaConnectionSecCertificate,
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			connection.BootstrapServers = &streamConnection.Spec.KafkaConfig.BootstrapServers
+			connection.Authentication = &admin.StreamsKafkaAuthentication{
+				Mechanism: &streamConnection.Spec.KafkaConfig.Authentication.Mechanism,
+				Username:  pointer.MakePtr(authData[kafkaConnectionAuthUsername]),
+				Password:  pointer.MakePtr(authData[kafkaConnectionAuthPassword]),
+			}
+			connection.Security = &admin.StreamsKafkaSecurity{
+				BrokerPublicCertificate: pointer.MakePtr(secData[kafkaConnectionSecCertificate]),
+				Protocol:                &streamConnection.Spec.KafkaConfig.Security.Protocol,
+			}
+			connection.Config = &streamConnection.Spec.KafkaConfig.Config
+		}
+
+		return &connection, nil
+	}
+}
+
+func getSecretData(ctx context.Context, k8sClient client.Client, ref client.ObjectKey, keys ...string) (map[string]string, error) {
+	secret := corev1.Secret{}
+	err := k8sClient.Get(ctx, ref, &secret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve secret %v: %w", ref, err)
+	}
+
+	data := map[string]string{}
+	for _, key := range keys {
+		val, ok := secret.Data[key]
+		if !ok {
+			return nil, fmt.Errorf("key %s is not present in the secret %v", key, ref)
+		}
+
+		data[key] = string(val)
+	}
+
+	return data, nil
+}
+
+func hasStreamConnectionChanged(
+	streamConnection *akov2.AtlasStreamConnection,
+	atlasStreamConnection admin.StreamsConnection,
```

Comment:
let's make this a pointer, there is no need for a shallow copy here.

Reply:
As I mutate the struct in the function I believe is safer to pass a copy to avoid side effects (even though right now I only unset a no relevant field `Links`)

Reply:
the struct has nested pointers so shallow-copying is having no effect. there is no point in passing a value here.

Reply:
if you want to be safe, you need to do a deep copy of `admin.StreamsConnection` which I believe the SDK has no methods for.